src/benchmarking/randomized_benchmarking.jl

```diff
@@ -0,0 +1,342 @@
+- `num_bits_on_device::Int = 0`: the total number of classical bits on the device.
+- `target_qubits::Array{Int}`: a list of the qubits on which benchmarking is performed. By default, all qubits on the device are benchmarked.
+- `fit_properties`: the [`RandomizedBenchmarkingFitProperties`](@ref) which define the model for fitting the averaged sequence fidelity.
+- `sequence_length_list::Array{Int}`: a list indicating the number of Clifford operations in each circuit. For instance, specifying [1, 2] will lead to the generation of multiple circuits with 1 Clifford operation and multiple circuits with 2 Clifford operations.
+- `num_circuits_per_length::Array{Int}`: a list defining how many circuits to generate for each sequence length. By default, 100 circuits are generated for each sequence length.
+"""
+@with_kw struct RandomizedBenchmarkingProperties
+    num_qubits_on_device::Int; @assert num_qubits_on_device > 0
+    num_bits_on_device::Int = 0; @assert num_bits_on_device >= 0
+    target_qubits::Array{Int} = 1:num_qubits_on_device
+    fit_properties::RandomizedBenchmarkingFitProperties =
+        RandomizedBenchmarkingFitProperties(0)
+
+    sequence_length_list::Array{Int}; @assert all(x->(x>0), sequence_length_list)
+    
+    num_circuits_per_length::Array{Int} = 100*ones(Int, length(sequence_length_list));
+        @assert all(x->(x>0), num_circuits_per_length)
+        @assert length(sequence_length_list) == length(num_circuits_per_length)
+end
+
+"""
+    RandomizedBenchmarkingResults
+
+The results of a randomized benchmarking calculation.
+
+# Fields
+- `sequence_length_list`: a list of the sequence lengths for which randomized circuits will be generated.
+- `sequence_fidelities`: a list giving the average fidelity for each sequence length.
+- `fit_results`: a fit of the averaged sequence fidelity, stored as [`RandomizedBenchmarkingFitResults`](@ref).
+- `average_clifford_fidelity`: the average fidelity of each Clifford operation.
+"""
+struct RandomizedBenchmarkingResults
+    sequence_length_list
+    sequence_fidelities
+    fit_results::RandomizedBenchmarkingFitResults
+    average_clifford_fidelity
+end
+
+"""
+    run_randomized_benchmarking(simulate_shots, properties::RandomizedBenchmarkingProperties, transpile! = (f(x)=x))
+
+Conducts randomized benchmarking and returns [`RandomizedBenchmarkingResults`](@ref).
+
+# Arguments
+- `simulate_shots`: a function which takes in an array of `QuantumCircuit` and returns an array of bit strings. The function should not transpile the circuits since it could transpile all the gates into a single identity operation. Use the transpile! argument instead.
+- `properties`: the [`RandomizedBenchmarkingProperties`](@ref) influencing benchmarking such as the list of target qubits and the list of sequence lengths.
+- `transpile!`: a function which transpiles a single circuit. It transpiles every Clifford operation individually.
+"""
+function run_randomized_benchmarking(simulate_shots,
```

Comment:
What is the data type of similate_shots? Please specify in the function declaration. Don't simply rely on documentation.